internal/app/starter/master_linux.go

```diff
@@ -11,129 +11,131 @@ import (
 	"net"
 	"os"
 	"os/signal"
-	"runtime"
 	"syscall"
-	"time"
 
 	"github.com/sylabs/singularity/internal/pkg/runtime/engines/engine"
 	"github.com/sylabs/singularity/internal/pkg/sylog"
 	"github.com/sylabs/singularity/internal/pkg/util/mainthread"
 )
 
-// Master initializes a runtime engine and runs it
-func Master(rpcSocket, masterSocket int, isInstance bool, containerPid int, e *engine.Engine) {
-	var fatal error
-	var status syscall.WaitStatus
-	fatalChan := make(chan error, 1)
-
-	// we could receive signal from child with CreateContainer call so we
-	// set the signal handler earlier to queue signals until MonitorContainer
-	// is called to handle them
-	signals := make(chan os.Signal, 1)
-	signal.Notify(signals)
+func createContainer(rpcSocket int, containerPid int, e *engine.Engine, fatalChan chan error) {
+	comm := os.NewFile(uintptr(rpcSocket), "rpc-socket")
+	if comm == nil {
+		fatalChan <- fmt.Errorf("bad RPC socket file descriptor")
+		return
+	}
+	rpcConn, err := net.FileConn(comm)
+	comm.Close()
+	if err != nil {
+		fatalChan <- fmt.Errorf("failed to copy unix socket descriptor: %s", err)
+		return
+	}
 
-	ppid := os.Getppid()
+	err = e.CreateContainer(containerPid, rpcConn)
+	if err != nil {
+		fatalChan <- fmt.Errorf("container creation failed: %s", err)
+		return
+	}
 
-	go func() {
-		comm := os.NewFile(uintptr(rpcSocket), "socket")
-		rpcConn, err := net.FileConn(comm)
-		comm.Close()
-		if err != nil {
-			fatalChan <- fmt.Errorf("failed to copy unix socket descriptor: %s", err)
-			return
-		}
+	rpcConn.Close()
```

Comment:
Should this call be in defer?

Reply:
Agreed. Should this be a deferred action?

Reply:
The connection to RPC is closed only when container creation succeed in order to not execute stage 2 process (container process) in case of failure, then the master process will automatically killed the RPC process at exit